File: MinecraftManhunt/src/main/java/com/yoonicode/minecraftmanhuntplus/PluginCommands.java
package com.yoonicode.minecraftmanhuntplus;

import org.bukkit.*;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.HumanEntity;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.bukkit.scheduler.BukkitScheduler;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class PluginCommands implements CommandExecutor {

    public static final String[] registeredCommands = {
            "hunter",
            "runner",
            "spectator",
            "clearteams",
            "start",
            "end",
            "compass",
            "setheadstart",
            "runnerhelp",
            "hunterhelp",
            "extradrops",
            "chestgenerate",
            "hasteboost",
            "allhelp"
    };
    public boolean hitHasRegistered; // used for startGameByHit option
    public boolean extraDrops;
    public boolean chestGenerate;

    int compassTask = -1;
    int dangerLevelTask = -1;
    public boolean gameIsRunning = false;

    boolean worldBorderModified;
    private final PluginMain main;
    public boolean runnerHelp = false;
    public boolean hunterHelp = false;
    public boolean hasteBoost = false;
    private boolean cutClean = false;

    public PluginCommands(PluginMain main) {
        this.main = main;
    }



    public List<String> getCompletions(String[] args, List<String> existingCompletions){
        switch (args[0]){
            case "/hunter":
            case "/runner":
            case "/spectator": {
                return Bukkit.getOnlinePlayers().stream().map(HumanEntity::getName).collect(Collectors.toList());
            }
            case "/start":
            case "/end":
            case "/chestgenerate":
            case "/compass":
            case "/clearteams":
            case "/hasteboost":
            case "/allhelp":
                return new ArrayList<String>();
            case "/setheadstart": {
                return new ArrayList<String>(){
                    {
                        add("0");
                        add("30");
                        add("60");
                    }
                };
            }
            default:
                return existingCompletions;

        }
    }

    public boolean onCommand(CommandSender commandSender, Command command, String label, String[] args) {
        if ("hunter".equals(label)) {
            if (args.length != 1) return false;
            Player target = Bukkit.getPlayer(args[0]);
            if (target == null) {
                commandSender.sendMessage("Target is not online");
                return false;
            }
            if (isOnTargetTeam(commandSender, target, main.hunters)) return true;
            main.runners.remove(target);
            main.spectators.remove(target);
            main.hunters.add(target);
            main.hunterDeaths.put(target,0);
            Bukkit.broadcastMessage(target.getName() + " is now a " + ChatColor.RED + ChatColor.BOLD + "HUNTER!");
            return true;
        }
        else if ("runner".equals(label)) {
            if (args.length != 1) return false;
            Player target = Bukkit.getPlayer(args[0]);
            if (target == null) {
                commandSender.sendMessage("Target is not online");
                return false;
            }
            if (isOnTargetTeam(commandSender, target, main.runners)) return true;
            main.hunters.remove(target);
            main.spectators.remove(target);
            main.runners.add(target);
            main.runnerDeaths.put(target,0);

            Bukkit.broadcastMessage(target.getName() + " is now a " + ChatColor.GREEN + ChatColor.BOLD + "RUNNER!");
            return true;
        }
        else if ("spectator".equals(label)) {
            if (args.length != 1) return false;
            Player target = Bukkit.getPlayer(args[0]);
            if (target == null) {
                commandSender.sendMessage("Target is not online");
                return false;
            }
            if (isOnTargetTeam(commandSender, target, main.spectators)) return true;
            main.runners.remove(target);
            main.hunters.remove(target);
            main.spectators.add(target);
            target.sendMessage("You have been marked as a spectator.");
            commandSender.sendMessage("Marked player as spectator");
            return true;
        }
        else if ("start".equals(label)) {
            if(gameIsRunning){
                commandSender.sendMessage("Game is already in progress. Use /end before starting another game.");
                return true;
            }
            if (main.runners.size() < 1 && !main.debugMode) {
                commandSender.sendMessage("Not enough speedrunners to start");
                return true;
            }
            commandSender.sendMessage("Starting game...");
            main.targets.clear();
            int headStartDuration = main.getConfig().getInt("headStartDuration");
            if (main.getConfig().getBoolean("clearItemDropsOnStart", false)) {
                commandSender.getServer().dispatchCommand(Bukkit.getConsoleSender(), "minecraft:kill @e[type=item]");
            }
            updateWorld();
            updateSpectators();
            runnersState();
            huntersState(headStartDuration);
            compassTask = Bukkit.getScheduler().scheduleSyncRepeatingTask(main, new Runnable() {
                public void run() {
                    main.getTaskManager().updateCompass();
                }
            }, 0L, 20L);

            if (hasteBoost) {
                Bukkit.getScheduler().scheduleSyncRepeatingTask(main, new Runnable() {
                    public void run() {
                        main.getTaskManager().giveHaste();
                    }
                }, 460L, 1200L);
            }

            //todo: How often does this actually need to repeat?
            Bukkit.getScheduler().scheduleSyncRepeatingTask(main, new Runnable() {
                @Override
                public void run() {
                    main.getTaskManager().showGlow();
                }
            }, 0,600); // 2700? More?
            gameIsRunning = true;
            sendStartMessage();
            return true;
        }
        else if ("end".equals(label)) {
            if(!gameIsRunning){
                commandSender.sendMessage("There is no game in progress. Use /start to start a new game.");
                return true;
            }
            if(main.getConfig().getBoolean("startGameByHit", false)){
                hitHasRegistered = false;
            }
            BukkitScheduler scheduler = Bukkit.getScheduler();
            if (compassTask != -1) {
                scheduler.cancelTask(compassTask);
                compassTask = -1;
            }
            if (dangerLevelTask != -1) {
                scheduler.cancelTask(dangerLevelTask);
                dangerLevelTask = -1;
            }
            worldBorderModified = false;
            Bukkit.broadcastMessage("Manhunt stopped!");
            gameIsRunning = false;
            return true;
        }
        else if("compass".equals(label)) {
            Player sender = (Player) commandSender;
            sender.getInventory().addItem(new ItemStack(Material.COMPASS, 1));
            commandSender.sendMessage("Here you go!");
            return true;
        }
        else if("clearteams".equals(label)) {
            commandSender.sendMessage(clearTeams());
            return true;
        }
        else if("setheadstart".equals(label)) {
            main.logger.info("setheadstart called.");
            if(args.length == 0){
                commandSender.sendMessage("Provide a headstart duration as a nonnegative integer");
                return false;
            }
            int duration;
            try {
                duration = Integer.parseInt(args[0]);
            } catch(NumberFormatException e){
                commandSender.sendMessage("Headstart duration must be a nonnegative integer");
                return false;
            }
            if(duration < 0){
                commandSender.sendMessage("Headstart duration must be greater than or equal to 0");
                return false;
            }
            main.getConfig().set("headStartDuration", duration);
            main.saveConfig();
            commandSender.sendMessage("Headstart set to " + duration);
            return true;
        }
        else if("runnerhelp".equals(label)){
            if (gameIsRunning){
                commandSender.sendMessage("Game is already in progress. Restart a game to change this option");
                return true;
            }
            if (args.length != 0){
                commandSender.sendMessage("Illegal format. Use /chestgenerate.");
                return true;
            }
            runnerHelp=!runnerHelp;
            commandSender.sendMessage("Runner help is set to: " + runnerHelp);
        }
        else if("hunterhelp".equals(label)){
            if (gameIsRunning){
                commandSender.sendMessage("Game is already in progress. Restart a game to change this option");
                return true;
            }
            if (args.length != 0){
                commandSender.sendMessage("Illegal format. Use /chestgenerate.");
                return true;
            }
            hunterHelp=!hunterHelp;
            commandSender.sendMessage("Hunter help is set to: " + hunterHelp);


        }
        else if("extradrops".equals(label)){
            if (gameIsRunning){
                commandSender.sendMessage("Game is already in progress. Restart a game to change this option");
                return true;
            }
            if (args.length != 0){
                commandSender.sendMessage("Illegal format. Use /chestgenerate.");
                return true;
            }
            extraDrops=!extraDrops;
            commandSender.sendMessage("Extra drops is set to: " + extraDrops);
        }
        else if ("chestgenerate".equals(label)){
            if (gameIsRunning){
                commandSender.sendMessage("Game is already in progress. Restart a game to change this option");
                return true;
            }
            if (args.length != 0){
                commandSender.sendMessage("Illegal format. Use /chestgenerate.");
                return true;
            }
            chestGenerate = !chestGenerate;
            commandSender.sendMessage("Random Chest spawns is set to: " + chestGenerate);
            return true;
        }
        else if ("hasteboost".equals(label)){
            if (gameIsRunning){
                commandSender.sendMessage("Game is already in progress. Restart a game to change this option");
                return true;
            }
            if (args.length != 0){
                commandSender.sendMessage("Illegal format. Use /chestgenerate.");
                return true;
            }
            hasteBoost = !hasteBoost;
            commandSender.sendMessage("Haste boost is set to: " + hasteBoost);
            return true;
        }
        else if ("allhelp".equals(label)){
            if (gameIsRunning){
                commandSender.sendMessage("Game is already in progress. Restart a game to change this option");
                return true;
            }
            if (args.length != 0){
                commandSender.sendMessage("Illegal format. Use /chestgenerate.");
                return true;
            }
            runnerHelp = true;
            hunterHelp = true;
            extraDrops = true;
            chestGenerate = true;
            hasteBoost = true;
            cutClean = true;
            commandSender.sendMessage("All helper methods are enabled");
            return true;
        }
        else if ("cutclean".equals(label)){
            if (gameIsRunning){
                commandSender.sendMessage("Game is already in progress. Restart a game to change this option");
            }
            if (args.length != 0){
                commandSender.sendMessage("Illegal format. Use /chestgenerate.");
                return true;
            }
            cutClean = !cutClean;
            commandSender.sendMessage("Cut Clean is set to: " + cutClean);
            return true;
        }
        return false;

    }

    private boolean isOnTargetTeam(CommandSender commandSender, Player target, ArrayList<Player> team) {
        if (team.stream().anyMatch(player -> player.getName().equals(target.getName()))){
            commandSender.sendMessage("Target is already on this team");
            return true;
        }
        return false;
    }

    private void sendStartMessage() {
        Bukkit.broadcastMessage(
                ChatColor.DARK_RED.toString() + ChatColor.UNDERLINE + "Berner er gay!" + ChatColor.RESET + "\n" +
                        ChatColor.AQUA + "Tarun " + ChatColor.GOLD + "Tarun" + ChatColor.AQUA + " Tarun" + "\n" +
                        ChatColor.GREEN + "TarunTårnet! " + ChatColor.GOLD + "TARUNSAN! TARUNSAN! TARUNSAN!" + ChatColor.GREEN + " BI is love, BI is life -Tarun, probably" + "\n" +
                        ChatColor.UNDERLINE + ChatColor.BOLD + ChatColor.LIGHT_PURPLE + "MANHUNT STARTED!" + ChatColor.RESET + ChatColor.DARK_GRAY + " NTarunNU"
        );
    }

    private void huntersState(int headStartDuration) {
        for (Player player : main.hunters) {
            if (player == null) continue;
            player.setGameMode(GameMode.SURVIVAL);
            player.addPotionEffect(new PotionEffect(PotionEffectType.SLOW, 20 * headStartDuration, 5));
            player.addPotionEffect(new PotionEffect(PotionEffectType.BLINDNESS, 20 * headStartDuration, 3));
            player.addPotionEffect(new PotionEffect(PotionEffectType.SLOW_DIGGING, 20 * headStartDuration, 10));
            player.setHealth(20.0);
            player.setFoodLevel(20);

            if (main.getConfig().getBoolean("clearHunterInvOnStart", false)) {
                player.getInventory().clear();
                player.setExp(0);
                player.setLevel(0);
            }

            player.getInventory().addItem(new ItemStack(Material.COMPASS, 1));

        }
    }

    private void runnersState() {
        for (Player player : main.runners) {
            if (player == null) continue;
            player.setGameMode(GameMode.SURVIVAL);
            player.setHealthScale(20.0);
            player.setMaxHealth(20.0);
            player.setHealth(20.0);
            player.setFoodLevel(20);
            player.addPotionEffect(new PotionEffect(PotionEffectType.SPEED, 450,1));
            player.addPotionEffect(new PotionEffect(PotionEffectType.FAST_DIGGING, Integer.MAX_VALUE,1));

            if (main.getConfig().getBoolean("clearRunnerInvOnStart", false)) {
                player.getInventory().clear();
                player.setExp(0);
                player.setLevel(0);
            }

        }
    }

    private void updateSpectators() {
        for (Player player : main.spectators) {
            if (player == null) continue;
            player.setGameMode(GameMode.SPECTATOR);
        }
    }

    private void updateWorld() {
        if (worldBorderModified) {
            WorldBorder wb = main.world.getWorldBorder();
            wb.setCenter(0.5, 0.5);
            wb.setSize(60000000);
        }

        List<World> worlds = Bukkit.getWorlds();
        World world1 = worlds.get(0);
        if (main.getConfig().getBoolean("setTimeToZero", true)) {
            main.world = world1;
            main.getWorld().setTime(0);
        }
    }

    public String clearTeams(){
        int playersCleared = main.hunters.size() + main.spectators.size() + main.runners.size();
        main.hunters.clear();
        main.spectators.clear();
        main.runners.clear();
        return playersCleared + " players cleared from teams";
    }

    public boolean isCutClean() {
        return cutClean;
    }
}

File: MinecraftManhunt/src/main/java/manhunt_plus/PluginCommands.kt
package manhunt_plus

import manhunt_plus.SessionHandler.start
import manhunt_plus.SessionHandler.end
import manhunt_plus.SessionHandler.pause
import manhunt_plus.SessionHandler.unPause
import org.bukkit.command.CommandExecutor
import org.bukkit.entity.HumanEntity
import java.util.stream.Collectors
import org.bukkit.entity.Player
import org.bukkit.*
import org.bukkit.inventory.ItemStack
import java.lang.NumberFormatException
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.inventory.meta.PotionMeta
import org.bukkit.potion.PotionData
import org.bukkit.potion.PotionType

/**
 * The Commands used in Manhunt
 *
 * @property main main used.
 */
class PluginCommands(private val main: PluginMain) : CommandExecutor {
    private var glow: Boolean = true;
    private var supplyDrops: Boolean = true;
    private var hitHasRegistered // used for startGameByHit option
            = false
    @JvmField
    var extraDrops = true
    @JvmField
    var chestGenerate = true
    @JvmField
    var compassTask = -1
    private var dangerLevelTask = -1
    @JvmField
    var gameIsRunning = false
    private var worldBorderModified = false
    @JvmField
    var runnerHelp = true
    @JvmField
    var hunterHelp = true
    private var hasteBoost = true
    var isCutClean = true
        private set

    /**
     * Completes the command
     *
     * @param args the command being written
     * @param existingCompletions the completions for the command
     * @return the completions
     */
    fun getCompletions(args: Array<String?>, existingCompletions: List<String>): List<String> {
        return when (args[0]) {
            "/hunter", "/runner", "/spectator" -> {
                Bukkit.getOnlinePlayers().stream().map { obj: HumanEntity -> obj.name }.collect(Collectors.toList())
            }
            "/start", "/end", "/chestgenerate", "/compass", "/clearteams", "/hasteboost", "/allhelp", "/cutclean", "/pause", "/unpause", "/supplydrops" -> mutableListOf()
            "/setheadstart" -> {
                mutableListOf("0", "30", "60")

            }
            "/health" -> {
                mutableListOf("20", "40")
            }
            "glow" -> {
                mutableListOf("true", "false")
            }
            else -> existingCompletions
        }
    }

    /**
     * When a command is run inside minecraft
     *
     * @param commandSender the commandSender sending the command (is a player)
     * @param command I have no idea
     * @param label the first part that comes after the "/". E.g., "/hunter" will have label hunter
     * @param args a list of after the first part of the command.
     * @return a boolean for whether the command worked or not.
     */
    override fun onCommand(commandSender: CommandSender, command: Command, label: String, args: Array<String>): Boolean {
        when (label) {
            "hunter" -> {
                if (args.size != 1) return false
                val target = Bukkit.getPlayer(args[0])
                if (target == null) {
                    commandSender.sendMessage("Target is not online")
                    return false
                }
                if (isOnTargetTeam(commandSender, target, main.hunters)) return true
                main.runners.remove(target)
                main.spectators.remove(target)
                main.hunters.add(target)
                main.hunterDeaths[target] = 0
                Bukkit.broadcastMessage(target.name + " is now a " + ChatColor.RED + ChatColor.BOLD + "HUNTER!")
                return true
            }
            "runner" -> {
                if (args.size != 1) return false
                val target = Bukkit.getPlayer(args[0])
                if (target == null) {
                    commandSender.sendMessage("Target is not online")
                    return false
                }
                if (isOnTargetTeam(commandSender, target, main.runners)) return true
                main.hunters.remove(target)
                main.spectators.remove(target)
                main.runners.add(target)
                main.runnerDeaths[target] = 0
                Bukkit.broadcastMessage(target.name + " is now a " + ChatColor.GREEN + ChatColor.BOLD + "RUNNER!")
                return true
            }
            "spectator" -> {
                if (args.size != 1) return false
                val target = Bukkit.getPlayer(args[0])
                if (target == null) {
                    commandSender.sendMessage("Target is not online")
                    return false
                }
                if (isOnTargetTeam(commandSender, target, main.spectators)) return true
                main.runners.remove(target)
                main.hunters.remove(target)
                main.spectators.add(target)
                target.sendMessage("You have been marked as a spectator.")
                commandSender.sendMessage("Marked player as spectator")
                return true
            }
            "start" -> {
//            if (!main.getConfig().getBoolean("gameIsOver")){
//                commandSender.sendMessage("This game is paused, do not use /start. Use /unpause to continue.");
//                return true;
//            }
                if (gameIsRunning) {
                    commandSender.sendMessage("Game is already in progress. Use /end before starting another game.")
                    return true
                }
                if (main.runners.size < 1 && !main.debugMode) {
                    commandSender.sendMessage("Not enough speedrunners to start")
                    return true
                }
                commandSender.sendMessage("Starting game...")
                main.targets.clear()
                val headStartDuration = main.config.getInt("headStartDuration")
                if (main.config.getBoolean("clearItemDropsOnStart", false)) {
                    commandSender.server.dispatchCommand(Bukkit.getConsoleSender(), "minecraft:kill @e[type=item]")
                }
                updateWorld()
                updateSpectators()
                runnersState()
                huntersState(headStartDuration)
                //            teleportPlayers();
                compassTask = Bukkit.getScheduler().scheduleSyncRepeatingTask(main, { main.taskManager.updateCompass() }, 0L, 20L)
                if (hasteBoost) {
                    Bukkit.getScheduler().scheduleSyncRepeatingTask(main, { main.taskManager.giveHaste() }, 460L, 1200L)
                }
                if (supplyDrops){
                    Bukkit.getScheduler().scheduleSyncRepeatingTask(main,{
                        Bukkit.broadcastMessage("Supply Drop lands in 3 minutes, be ready for a chance at great loot!")
                    },20400L,24000L)
                    Bukkit.getScheduler().scheduleSyncRepeatingTask(main, {
                        main.taskManager.supplyDrop()
                    },24000L, 24000L)
                }

                if (glow) {
                    Bukkit.getScheduler().scheduleSyncDelayedTask(main, {
                        main.glowHandler.showGlow()
                    }, 20L)
                }

                gameIsRunning = true
                sendStartMessage()
                start(main)
                return true
            }
            "end" -> {
                if (!gameIsRunning) {
                    commandSender.sendMessage("There is no game in progress. Use /start to start a new game.")
                    return true
                }
                if (main.config.getBoolean("startGameByHit", false)) {
                    hitHasRegistered = false
                }
                val scheduler = Bukkit.getScheduler()
                if (compassTask != -1) {
                    scheduler.cancelTask(compassTask)
                    compassTask = -1
                }
                if (dangerLevelTask != -1) {
                    scheduler.cancelTask(dangerLevelTask)
                    dangerLevelTask = -1
                }
                worldBorderModified = false
                Bukkit.broadcastMessage("Manhunt stopped!")
                gameIsRunning = false
                end(main)
                return true
            }
            "compass" -> {
                if (args.isNotEmpty()) {
                    commandSender.sendMessage("Illegal format, please use /compass")
                }
                val sender = commandSender as Player
                sender.inventory.addItem(ItemStack(Material.COMPASS, 1))
                commandSender.sendMessage("Here you go!")
                return true
            }
            "clearteams" -> {
                commandSender.sendMessage(clearTeams())
                return true
            }
            "setheadstart" -> {
                main.mainLogger.info("setheadstart called.")
                if (args.isEmpty()) {
                    commandSender.sendMessage("Provide a headstart duration as a nonnegative integer")
                    return false
                }
                val duration: Int = try {
                    args[0].toInt()
                } catch (e: NumberFormatException) {
                    commandSender.sendMessage("Headstart duration must be a nonnegative integer")
                    return false
                }
                if (duration < 0) {
                    commandSender.sendMessage("Headstart duration must be greater than or equal to 0")
                    return false
                }
                main.config["headStartDuration"] = duration
                main.saveConfig()
                commandSender.sendMessage("Headstart set to $duration")
                return true
            }
            "runnerhelp" -> {
                if (illegalCommandCall(commandSender, args, "runnerhelp")) return true
                runnerHelp = !runnerHelp
                commandSender.sendMessage("Runner help is set to: $runnerHelp")
            }
            "hunterhelp" -> {
                if (illegalCommandCall(commandSender, args, "hunterhelp")) return true
                hunterHelp = !hunterHelp
                commandSender.sendMessage("Hunter help is set to: $hunterHelp")
            }
            "extradrops" -> {
                if (illegalCommandCall(commandSender, args, "extradrops")) return true
                extraDrops = !extraDrops
                commandSender.sendMessage("Extra drops is set to: $extraDrops")
            }
            "chestgenerate" -> {
                if (illegalCommandCall(commandSender, args, "chestgenerate")) return true
                chestGenerate = !chestGenerate
                commandSender.sendMessage("Random Chest spawns is set to: $chestGenerate")
                return true
            }
            "hasteboost" -> {
                if (illegalCommandCall(commandSender, args, "hasteboost")) return true
                hasteBoost = !hasteBoost
                commandSender.sendMessage("Haste boost is set to: $hasteBoost")
                return true
            }
            "allhelp" -> {
                if (illegalCommandCall(commandSender, args, "allhelp")) return true
                runnerHelp = true
                hunterHelp = true
                extraDrops = true
                chestGenerate = true
                hasteBoost = true
                isCutClean = true
                main.health = 40.0
                supplyDrops = true
                glow = true
                commandSender.sendMessage("All helper methods are enabled")
                return true
            }
            "cutclean" -> {
                if (illegalCommandCall(commandSender, args, "cutclean")) return true
                isCutClean = !isCutClean
                commandSender.sendMessage("Cut Clean is set to: $isCutClean")
                return true
            }
            "pause" -> {
                if (args.isNotEmpty()) {
                    return true
                }
                pause(main)
            }
            "unpause" -> {
                if (args.isNotEmpty()) {
                    return true
                }
                unPause(main)
            }
            "health" -> {
                if (gameInSession(commandSender)) return true
                if (args.isEmpty()) {
                    commandSender.sendMessage("Must provide health value (number of half-hearts. Standard is 20)")
                    return true
                }
                else if (args.size > 1){
                    commandSender.sendMessage("Invalid format. Use /health <number>")
                    return true
                }
                try {
                    main.health = args[0].toDouble()
                } catch (e: NumberFormatException){
                    commandSender.sendMessage("Please provide a valid number")
                }
            }
            "supplydrops" -> {
                if (illegalCommandCall(commandSender,args, "supplydrops" )) return true
                supplyDrops = !supplyDrops
                commandSender.sendMessage("Supply drops is set to: $supplyDrops")
                return true
            }
            "glow" -> {
                if (gameIsRunning){
                    commandSender.sendMessage("Game is running, end game before using this command.")
                    return true
                }
                if (args.size != 1){
                    glow = !glow
                    commandSender.sendMessage("Glow is now set to $glow")
                    return true
                }
                if (args[0].equals("true",true)){
                    this.glow = true;
                    commandSender.sendMessage("Glow is now set to $glow")
                    return true
                }
                else if (args[0].equals("false",true)){
                    this.glow = false
                    commandSender.sendMessage("Glow is now set to $glow")
                    return true
                }
            }
        }
        return false
    }

    //    private void teleportPlayers() {
    //        for (Player player : Bukkit.getOnlinePlayers()){
    //            player.teleport(player.getWorld().getSpawnLocation());
    //        }
    //    }

    /**
     * Checks if the command call is illegal
     *
     * @param commandSender the sender of the command
     * @param args a list of after the first part of the command.
     * @param command The command being sent
     * @return a boolean for whether the call is illegal or not
     */
    private fun illegalCommandCall(commandSender: CommandSender, args: Array<String>, command: String): Boolean {
        if (gameInSession(commandSender)) return true
        if (args.isNotEmpty()) {
            commandSender.sendMessage("Illegal format. Use /$command.")
            return true
        }
        return false
    }

    /**
     * Check if the game is running
     *
     * @param commandSender the sender of the command.
     */
    private fun gameInSession(commandSender: CommandSender) : Boolean {
        if (gameIsRunning) {
            commandSender.sendMessage("Game is already in progress. Restart a game to change this option")
            return true
        }
        return false
    }

    /**
     * Check if the target of the command is on the desired team already.
     *
     * @param commandSender the sender of the command.
     * @param target the target of the command.
     * @param team the target team.
     * @return true if the target is on that team.
     */
    private fun isOnTargetTeam(commandSender: CommandSender, target: Player, team: List<Player>): Boolean {
        if (team.stream().anyMatch { player: Player -> player.name == target.name }) {
            commandSender.sendMessage("Target is already on this team")
            return true
        }
        return false
    }

    /**
     * Send the message on start of a manhunt game
     */
    private fun sendStartMessage() {
        Bukkit.broadcastMessage(
                """${ChatColor.DARK_RED}${ChatColor.UNDERLINE}Berner er gay!${ChatColor.RESET}
                    ${ChatColor.AQUA}Tarun ${ChatColor.GOLD}Tarun${ChatColor.AQUA} Tarun
                    ${ChatColor.GREEN}TarunTårnet! ${ChatColor.GOLD}TARUNSAN! TARUNSAN! TARUNSAN!${ChatColor.GREEN} BI is love, BI is life -Tarun, probably
                    ${ChatColor.UNDERLINE}${ChatColor.BOLD}${ChatColor.LIGHT_PURPLE}MANHUNT STARTED!${ChatColor.RESET}${ChatColor.DARK_GRAY} NTarunNU"""
        )
    }

    /**
     * The hunters state when the game is started.
     *
     * @param headStartDuration the duration that hunters must wait.
     */
    @Suppress("DEPRECATION")
    private fun huntersState(headStartDuration: Int) {
        for (player in main.hunters) {
            player.addPotionEffect(PotionEffect(PotionEffectType.SLOW, 20 * headStartDuration, 5))
            player.addPotionEffect(PotionEffect(PotionEffectType.BLINDNESS, 20 * headStartDuration, 3))
            player.addPotionEffect(PotionEffect(PotionEffectType.SLOW_DIGGING, 20 * headStartDuration, 10))
            startState(player)
            Bukkit.getScheduler().scheduleSyncRepeatingTask(main, {
                main.taskManager.updateActionBar(player, main.hunters)
            }, 10L, 20L)
        }

    }

    private fun startState(player: Player) {
        player.gameMode = GameMode.SURVIVAL
        player.healthScale = main.health
        player.maxHealth = main.health
        player.health = main.health
        player.foodLevel = 20
        player.inventory.clear()
        player.exp = 0f
        player.level = 0
        player.inventory.addItem(ItemStack(Material.COMPASS, 1))
    }

    /**
     * The runners state when the game is started.
     */
    @Suppress("DEPRECATION")
    private fun runnersState() {
        for (player in main.runners) {
            startState(player)
            player.foodLevel = 20
            player.addPotionEffect(PotionEffect(PotionEffectType.SPEED, 1200, 1))
            Bukkit.getScheduler().scheduleSyncRepeatingTask(main, {
                main.taskManager.updateActionBar(player, main.runners)
            }, 0L, 20L)
            val speedPotion = createSpeedPotion()
            player.inventory.addItem(speedPotion)
        }
    }

    private fun createSpeedPotion(): ItemStack {
        val speedPotion = ItemStack(Material.POTION, 1)
        val meta = speedPotion.itemMeta as PotionMeta?
        meta?.basePotionData = PotionData(PotionType.SPEED)
        speedPotion.itemMeta = meta
        return speedPotion
    }

    /**
     * Set gamemode for spectators
     */
    private fun updateSpectators() {
        for (player in main.spectators) {
            player.gameMode = GameMode.SPECTATOR
        }
    }

    /**
     * Update the world border and time
     */
    private fun updateWorld() {
        if (worldBorderModified) {
            val wb = main.world?.worldBorder
            wb?.setCenter(0.5, 0.5)
            wb?.size = 60000000.0
        }
        if (main.config.getBoolean("setTimeToZero", true)) {
            main.world?.time = 0
            main.world?.clearWeatherDuration = 199999999
        }
    }

    /**
     * Cleat the teams
     *
     * @return a string message to the players
     */
    private fun clearTeams(): String {
        val playersCleared = main.hunters.size + main.spectators.size + main.runners.size
        main.hunters.clear()
        main.spectators.clear()
        main.runners.clear()
        return "$playersCleared players cleared from teams"
    }

    companion object {
        @JvmField
        val registeredCommands = arrayOf(
            "hunter",
            "runner",
            "spectator",
            "clearteams",
            "start",
            "end",
            "compass",
            "setheadstart",
            "runnerhelp",
            "hunterhelp",
            "extradrops",
            "chestgenerate",
            "hasteboost",
            "allhelp",
            "cutclean",
            "pause",
            "unpause",
            "health",
            "supplydrops",
            "glow"
        )
    }
}